source/simulator/stokes_matrix_free.cc

```diff
@@ -1598,6 +1702,146 @@ namespace aspect
+                        = out.template get_additional_output<MaterialModel::MaterialModelDerivatives<dim> >();
+
+                      Assert(derivatives != nullptr,
+                             ExcMessage ("Error: The Newton method requires the material to "
+                                         "compute derivatives."));
+
+                      const double viscosity_derivative_wrt_pressure
+                        = derivatives->viscosity_derivative_wrt_pressure[q];
+
+                      const SymmetricTensor<2,dim> viscosity_derivative_wrt_strain_rate
+                        = derivatives->viscosity_derivative_wrt_strain_rate[q];
+
+                      // use the spd factor when the stabilization is PD or SPD
+                      const double alpha =  (sim.newton_handler->parameters.velocity_block_stabilization
+                                             & Newton::Parameters::Stabilization::PD)
+                                            != Newton::Parameters::Stabilization::none
+                                            ?
+                                            Utilities::compute_spd_factor<dim>(out.viscosities[q], strain_rate, viscosity_derivative_wrt_strain_rate,
+                                                                               sim.newton_handler->parameters.SPD_safety_factor)
+                                            :
+                                            1;
+
+                      active_viscosity_derivative_wrt_pressure_table(cell,q)[i]
+                        = viscosity_derivative_wrt_pressure * newton_derivative_scaling_factor;
+
+                      for (unsigned int m=0; m<dim; ++m)
+                        for (unsigned int n=0; n<dim; ++n)
+                          {
+                            // TODO: why can we not assign tensors directly?
+                            active_strain_rate_table(cell, q)[m][n][i]
+                              = strain_rate[m][n];
+
+                            active_viscosity_derivative_wrt_strain_rate_table(cell, q)[m][n][i]
+                              = derivatives->viscosity_derivative_wrt_strain_rate[q][m][n]
+                                * newton_derivative_scaling_factor * alpha;
+                          }
+                    }
+                }
+            }
+          bool is_symmetric =
+            ((sim.newton_handler->parameters.preconditioner_stabilization
+              & Newton::Parameters::Stabilization::symmetric)
+             != Newton::Parameters::Stabilization::none);
+
+          stokes_matrix.fill_Newton_cell_data(active_viscosity_derivative_wrt_pressure_table,
+                                              active_strain_rate_table,
+                                              active_viscosity_derivative_wrt_strain_rate_table,
+                                              is_symmetric);
+
+          if (sim.parameters.n_expensive_stokes_solver_steps > 0)
+            {
```

Comment:
I am surprised that you only need this for the expensive solve. Are you sure?

Reply:
I am not sure, just copy and paste, the preconditioner doesn't work for now.

Reply:
I will remove them for this PR.